clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -343,15 +343,34 @@ static void collectSYCLAttributes(Sema &S, FunctionDecl *FD,
   if (!FD->hasAttrs())
     return;
 
-  llvm::copy_if(FD->getAttrs(), std::back_inserter(Attrs), [](Attr *A) {
-    // FIXME: Make this list self-adapt as new SYCL attributes are added.
-    return isa<IntelReqdSubGroupSizeAttr, IntelNamedSubGroupSizeAttr,
-               ReqdWorkGroupSizeAttr, SYCLIntelKernelArgsRestrictAttr,
-               SYCLIntelNumSimdWorkItemsAttr,
-               SYCLIntelSchedulerTargetFmaxMhzAttr,
-               SYCLIntelMaxWorkGroupSizeAttr, SYCLIntelMaxGlobalWorkDimAttr,
-               SYCLIntelNoGlobalWorkOffsetAttr, SYCLSimdAttr>(A);
-  });
+  // Attributes that should be propagated from device functions to a kernel
+  // in SYCL 1.2.1.
+  if (S.getASTContext().getLangOpts().getSYCLVersion() <
+      LangOptions::SYCL_2020) {
+    llvm::copy_if(FD->getAttrs(), std::back_inserter(Attrs), [](Attr *A) {
+      // FIXME: Make this list self-adapt as new SYCL attributes are added.
+      return isa<SYCLIntelKernelArgsRestrictAttr, SYCLIntelNumSimdWorkItemsAttr,
+                 SYCLIntelSchedulerTargetFmaxMhzAttr, ReqdWorkGroupSizeAttr,
+                 SYCLIntelMaxWorkGroupSizeAttr, SYCLIntelMaxGlobalWorkDimAttr,
+                 IntelReqdSubGroupSizeAttr, SYCLSimdAttr,
+                 SYCLIntelNoGlobalWorkOffsetAttr>(A);
+    });
+  } else {
+    // Attributes that should not be propagated from device functions to a
```

Comment:
The `if` branch are the attributes that should be propagated and the `else` branch is the ones that should not? But the `else` branch still copies this list into `Attrs`? I feel like something must be wrong here.

Reply:
The `if` branch here is to propagate attributes from child functions to kernel in SYCL 1.2.1. It should also contain code to copy attributes (Which don't propagate in SYCL 1.2.1) applied directly on kernel functor or lambda (i.e. DirectlyCalled = true). I think this patch accidently deleted the latter code. 

The `else` branch handles only attributes directly applied on kernel functor/lambda since in SYCL2020 there is no propagation from child functions.

Reply:
So this function is doing the same collecting task for two different reasons that are related only because they require us to walk over the list of attributes to find ones that match for a specific need?

Reply:
>>It should also contain code to copy attributes (Which don't propagate in SYCL 1.2.1) applied directly on kernel functor or lambda (i.e. DirectlyCalled = true). I think this patch accidently deleted the latter code.

 Thanks @elizabethandrews for pointing this. Yes, I will restore this with new patch.